kindelia_core/src/hvm.rs

```diff
@@ -1683,49 +1683,68 @@ impl Runtime {
   }
 
   // Full-reduce a term.
-  // TODO: rewrite as stack based? not needed?
   pub fn normalize(&mut self, host: Loc, mana:u64, seen: &mut HashSet<RawCell>) -> Result<RawCell, RuntimeError> {
-    let term  = ask_lnk(self, host);
-    if seen.contains(&term) {
-      Ok(term)
+    enum StackItem {
+      Host(Loc),
+      Linker(Loc)
```

Comment:
I think both can be done in a single step, but I'm not going to optimize this now.